Спринт №12

LasvegasandFlorida opened the pull request.

Commit:
- Исправлено изменение иконок об ошибках в зависимости от Темной\Светлой темы в меню Поиска

faritowich commented on app/src/main/java/com/example/playlistmakerprod/SearchActivity.kt:
```diff
@@ -113,76 +105,66 @@ class SearchActivity : AppCompatActivity() {
         }
         inputEditText.addTextChangedListener(simpleTextWatcher)
     }
+    private fun showHideError(code: Int)
```
Лучше не завязываться на code, потому что не очевидно, что именно за число принимает этот метод. Вот эти 0, 205, которые передаются в аргументы - непонятно, откуда они. Лучше всё-таки передавать в аргументы метода не инт, а enum класс статуса поиска, как я рекомендовал в прошлом ревью.

Commit:
- Добавлено сохранение настроек Темной темы
- Добавлено сохранение истории поиска экрана Поиск

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/App.kt:
```diff
@@ -0,0 +1,24 @@
+import androidx.appcompat.app.AppCompatDelegate
+
+class App : Application() {
+private var darkTheme = false
```
Немного съехало форматирование, можно использовать горячие клавиши  ctrl + alt + L или cmd + opt + L,  чтобы автоматически поправило в файле

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchActivity.kt:
```diff
@@ -44,7 +42,10 @@ class SearchActivity : AppCompatActivity() {
     companion object {
         private var instance: SearchActivity? = null
-
+        const val ON_SUCCESS = 200
```
Enum подошел бы чуточку лучше, чем константы, так как это ограниченное множество вариантов кодов

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchActivity.kt:
```diff
@@ -58,34 +59,45 @@ class SearchActivity : AppCompatActivity() {
-
-    }
-
     override fun onCreate(savedInstanceState: Bundle?) {
```
Метод стал уже достаточно большим, можно его поделить логически на несколько, чтобы было легче ориентироваться в коде

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchActivity.kt:
```diff
@@ -58,34 +59,45 @@ class SearchActivity : AppCompatActivity() {
+        inputEditText.setOnFocusChangeListener { _, hasFocus ->
+            if (hasFocus && inputEditText.text.isEmpty()) showHideError(SHOW_HISTORY)
+            else
+            {
```
Здесь тоже съехало форматирование, по код стайлу котлина открывающиеся скобки должны быть на одной строке с командой

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchActivity.kt:
```diff
@@ -58,34 +59,45 @@ class SearchActivity : AppCompatActivity() {
+        }
         updateButton.setOnClickListener {
+            if (updateButton.text == getText(R.string.update))
             onUpdateConnClick()
```
Если блок ифа на новой строке, то его стоит обернуть в фигурные скобки. Про код стайл можно почитать на этой странице: https://kotlinlang.org/docs/coding-conventions.html#control-flow-statements

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchHistory.kt:
```diff
@@ -0,0 +1,29 @@
+
+class SearchHistory(private val sharedPrefs: SharedPreferences) {
+    fun getSongs(listOfSongs:ArrayList<Track>)
+    {
```
Кажется, что в твоем опыте присутствует программирование на C#, такое форматирование характерно для него)

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchHistory.kt:
```diff
@@ -0,0 +1,29 @@
+class SearchHistory(private val sharedPrefs: SharedPreferences) {
+    fun getSongs(listOfSongs:ArrayList<Track>)
+    {
+        val tracksPrefs = TracksPreferences()
```
Было бы оптимальнее создать TracksPreferences переменной на уровне класса, чтобы создавать его только один раз, а не каждый раз, когда получается список треков или добавляется трек

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchHistory.kt:
```diff
@@ -0,0 +1,29 @@
+    fun getSongs(listOfSongs:ArrayList<Track>)
+    {
+        val tracksPrefs = TracksPreferences()
+        listOfSongs.addAll(tracksPrefs.read(sharedPrefs)!!)
```
Очень небезопасно использовать оператор !!, это надо быть на 100 процентов быть уверенным, что никогда null там не будет. Самый безопасный вариант - использовать ?.  или проверять на null

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchHistory.kt:
```diff
@@ -0,0 +1,29 @@
+        val sameSong = listOfSongs.find{it.trackId == track.trackId}
+        if (sameSong!=null)
+            listOfSongs.remove(sameSong)
+        else if (listOfSongs.size==10)
```
Лучше всего выносить числа в константы, чтобы они не были [магическими числами](https://ru.wikipedia.org/wiki/%D0%9C%D0%B0%D0%B3%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B5_%D1%87%D0%B8%D1%81%D0%BB%D0%BE_(%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5))

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchHistory.kt:
```diff
@@ -0,0 +1,29 @@
+        if (sameSong!=null)
+            listOfSongs.remove(sameSong)
+        else if (listOfSongs.size==10)
+            listOfSongs.removeAt(listOfSongs.size-1)
```
Можно было бы использовать метод .removeLast(), который под капотом удаляет последний элемент

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/Track.kt:
```diff
@@ -2,4 +2,4 @@ package com.example.playlistmakerprod
 import com.google.gson.annotations.SerializedName
 
-data class Track( val trackName:String, val artistName: String, @SerializedName("trackTimeMillis") val trackTime: String, val artworkUrl100: String)
+data class Track(val trackId:Int, val trackName:String, val artistName: String, @SerializedName("trackTimeMillis") val trackTime: String, val artworkUrl100: String)
```
Здесь было бы хорошо переносить поля на новые строки, так как класс уже достаточно большой. Хоткей выше тоже это решит)

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/TracksPreferences.kt:
```diff
@@ -0,0 +1,22 @@
+
+
+
+
```
Лучше не оставлять слишком много пустых строк, они ухудшают восприятие кода

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/TracksPreferences.kt:
```diff
@@ -0,0 +1,22 @@
+
+
+class TracksPreferences {
+    fun read(sharedPreferences: SharedPreferences): ArrayList<Track>? {
```
Если в этом методе делается все, чтобы ни в коем случае не вернулся null, то следует убрать его из возвращаемого типа, чтобы минимизировать использование опасного оператора !!

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchActivity.kt:
```diff
@@ -106,83 +117,95 @@ class SearchActivity : AppCompatActivity() {
             override fun afterTextChanged(s: Editable?) {
             }
         }
+        songsAdapter.onItemClick = {track ->
```
Можно было бы onItemClick сделать параметров в конструкторе адаптера и передавать сразу при создание адаптера, тогда и поле это было бы не null

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SearchActivity.kt:
```diff
@@ -44,7 +42,10 @@ class SearchActivity : AppCompatActivity() {
     }
     companion object {
         private var instance: SearchActivity? = null
```
![image](https://github.com/LasvegasandFlorida/PlaylistMakerProd/assets/129777932/c66048c3-e9b3-4d11-9454-2fe14ad5885c)

Здесь происходит классическая утечка памяти, о которой даже предупреждает студия. В чем суть: companion object - это статический класс, если он будет держать ссылку на активити, то ее не соберет garbage collector при смерти активити (например, при смене конфигурации), что будет занимать память.

Ты используешь этот код во вью холдере, чтобы получить application context, Но это абсолютно не нужно, так как во вью холдере можно получить контекст и так, пример в следующем треде

Anna-Malinina replied:
Статья про утечки памяти: https://www.programmersought.com/article/97295000566/. Там же написано, как можно добавить к себе в приложение leak canary, которая будет сигнализировать об утечках. Чтобы поймать эту утечку, следует походить по экранам несколько раз, по поиску в том числе, свернуть приложение, подождать. Будут появляться пуши от канарейки, что есть утечки

Anna-Malinina commented on app/src/main/java/com/example/playlistmakerprod/SongsViewHolder.kt:
```diff
@@ -11,13 +11,17 @@ import com.bumptech.glide.load.resource.bitmap.RoundedCorners
+        onItemClick?.invoke(tracks[adapterPosition])
+    }
+}
     fun bind(model: Track) {
```
Как получить контекст во вью холдере: здесь мы имеем доступ к itemView, которое является View, а у любой вью есть контекст, получаем его так:` itemView.context`. Тогда никаких опасных утечек памяти не потребуется

Anna-Malinina replied:
Также совет как получить скругленные углы, используя только ресурсы, без самописных вспомогательных методов: `itemView.resources.getDimensionPixelSize(Здесь ид ресурса, в котором указан размер радиуса в dp. По макету он должен быть 2)`

Commit:
- companion object в SearchActivity заменен на Enum class
- исправлены стили текстовок в соответсвии с макетом
- исправлена ошибка с утечной памяти в SearchActivity
- убрано использование опсного оператора "!!" в SearchHistory

Commit:
- убран allCaps у кнопки UpdateButton на экране SearchActivity
- выставлено максимальное количество строк для поля с наименованием трека в songs_view.xml (был уверен, что было выставлено в предыдущем комите, видимо ctrl+z подвел)
- выставлен корректный шрифт для Исполнителя и времени в songs_view.xml
